contrib/windows/exe_path_wrapper.c

```diff
@@ -0,0 +1,81 @@
+#include <windows.h>
+#include <tchar.h>
+#include <stdio.h>
+#include <shlwapi.h>
+
+/* The maximum path length we'll allow */
+#define MAX_PATH_LEN 1920
+
+/* PATH_ENTRIES is our simulated RPATH, usually of the form 'TEXT("../path1"), TEXT("../path2"), ...' */
+#ifndef PATH_ENTRIES
+#define PATH_ENTRIES  TEXT("")
+#endif
+
+LPWSTR pathEntries[] = {
+    PATH_ENTRIES
+};
+
+/* JULIA_EXE_PATH is the relative path to julia.exe */
+#ifndef JULIA_EXE_PATH
+#define JULIA_EXE_PATH "../libexec/julia.exe"
+#endif
+
+int wmain(int argc, wchar_t *argv[], wchar_t *envp[]) {
+    // Determine absolute path to true julia.exe sitting in `libexec/`
+    WCHAR currFileDir[MAX_PATH_LEN];
+    WCHAR juliaPath[MAX_PATH_LEN];
+    if (!GetModuleFileName(NULL, currFileDir, MAX_PATH_LEN)) {
+        fprintf(stderr, "ERROR: GetModuleFileName() failed with code %d\n", GetLastError());
+        exit(1);
+    }
+    PathRemoveFileSpec(currFileDir);
+    PathCombine(juliaPath, currFileDir, TEXT(JULIA_EXE_PATH));
+
+    // On windows, we simulate RPATH by pushing onto PATH
+    LPWSTR pathVal = (LPWSTR) malloc(MAX_PATH_LEN*sizeof(WCHAR));
+    DWORD dwRet = GetEnvironmentVariable(TEXT("PATH"), pathVal, MAX_PATH_LEN);
+    DWORD numPathEntries = sizeof(pathEntries)/sizeof(pathEntries[0]);
+    if (dwRet == 0) {
+        // If we cannot get PATH, then our job is easy!
+        pathVal[0] = '\0';
+    }
+    else {
+        // Otherwise, we append, if we have enough space to:
+        DWORD currFileDirLen = wcslen(currFileDir);
+        DWORD totalPathLen = dwRet + 1 + currFileDirLen;
+        int env_idx;
+        for (env_idx = 0; env_idx < numPathEntries; ++env_idx) {
+            totalPathLen += 1 + currFileDirLen + 1 + wcslen(pathEntries[env_idx]);
+        }
+        if (MAX_PATH_LEN < totalPathLen) {
+            fprintf(stderr, "ERROR: Cannot append entries to PATH: not enough space in environment block.  Reduce size of PATH!");
+            exit(1);
+        }
+        lstrcat(pathVal, TEXT(";"));
+    }
+    // We always add the current directory (e.g. `bin/`) to PATH so that we can find e.g. libjulia.dll
```

Comment:
This seems like a bad idea to add to PATH, as it means when we try to launch other, unrelated programs, they may accidentally pick up our vendored copies of these files and fail, instead of loading their intended dll files.

Reply:
The JLL rewrite will cause the only files within that `PATH` to be `libjulia.dll`; all other dependencies will be in their own artifacts directories.

Do you think we should save the unaltered `PATH` within a separate environment variable (`JULIA_REAL_PATH`) then once the true `julia.exe` is launched, we can reset `PATH` to `JULIA_REAL_PATH`?  That would at least make this more transparent.